DOCS-1804: Add hard hat detection tutorial (#2616)
viamrobotics/docs: docs/tutorials/projects/helmet.md

videos: ["/tutorials/helmet/hardhat.webm", "/tutorials/helmet/hardhat.mp4"]
images: ["/tutorials/helmet/hardhat-data.png"]
videoAlt: "A man without a hard hat is detected and labeled as No-Hardhat. Then he puts on a hard hat and a bounding box labeled Hardhat appears. He gives a thumbs-up to the camera."
tags: ["tutorial"]
authors: [Jessamy Taylor]
languages: []
viamresources: ["camera", "data_manager", "mlmodel", "vision"]
level: "Intermediate"
date: "2024-03-05"
cost: 120
---

{{<imgproc src="/tutorials/helmet/ppe-hooks.png" resize="x300" declaredimensions=true alt="Hard hats and neon reflective vests on hooks." class="alignright" style="max-width: 350px">}}

We all know personal protective equipment (PPE) helps keep us safe, but sometimes we need a reminder to use it consistently.
Luckily, you can address this problem using Viam's integrated [data capture](/data/capture/), [computer vision](/ml/vision/), and [webhooks](/build/configure/#webhooks), along with a hard hat detection model.

Through this tutorial you will build a system to look out for you and your team, sending an email notification when someone isn't wearing a hard hat.

> {{<gif webm_src="/tutorials/helmet/hardhat.webm" mp4_src="/tutorials/helmet/hardhat.mp4" alt="A man without a hard hat is detected and labeled as No-Hardhat. Then he puts on a hard hat and a bounding box labeled Hardhat appears. He gives a thumbs-up to the camera." class="alignleft" max-width="250px">}}

Author: yes :)

First, you'll set up and test the computer vision functionality.

> Next, you'll set up data capture to pull images of people without hard hats into the cloud.

Author: Ah fair enough! I'll do this:
```suggestion
Next, you'll set up the data service to capture images of people without hard hats and sync them to the cloud.
```

You could mount your machine in a stationary location like on a pole, or you could mount it on a rover.
This tutorial covers the software side; you can get creative with the hardware.

Note that your machine must be connected to the internet for data sync and email notifications to work.

### Required software

- [`viam-server`](/get-started/installation/)
- [Ultralytics YOLOv8](https://docs.ultralytics.com/), integrated with Viam using the [YOLOv8 modular service](https://github.com/viam-labs/YOLOv8)
- [YOLOv8 Hard Hat Detection model](https://huggingface.co/keremberke/yolov8s-hard-hat-detection)
- [`objectfilter-camera`](https://github.com/felixreichenbach/objectfilter-camera) {{< glossary_tooltip term_id="modular-resource" text="modular resource" >}}
- [Google Cloud Functions](https://cloud.google.com/functions)
- Python 3.8
- [SendGrid](https://sendgrid.com/en-us)

## Set up your helmet monitor

Plug your webcam into your computer.

Make sure your computer (whether it's a personal computer or an SBC) is connected to adequate power, and turn it on.

> If you haven't already, [install `viam-server`](/get-started/installation/) on the computer you're using for this project, and set up the connection between your machine and the [Viam app](https://app.viam.com).

Author: Okay I'm happy to point to the Viam app.

As for the why they should install `viam-server`, we don't usually do that? If they've gotten this far perhaps they know what Viam is about? It's sort of just a prerequisite, and the filtered-camera tutorial doesn't seem to say why you should install it either. Not sure how to sum it up without linking to Viam in 3 minutes etc.

I could change the heading of this section to "Prerequisites" if you think that would help?

> If you haven't already, [install `viam-server`](/get-started/installation/) on the computer you're using for this project, and set up the connection between your machine and the [Viam app](https://app.viam.com).

Reviewer: Not sure I understand - you'd prefer to just point them to create a new machine and follow the setup instructions until the machine is connected? I'm ok with that, just make sure you make it clear when they can return to the instructions here

## Configure the camera and computer vision

### Configure your physical camera

> Configure your [webcam](/components/camera/webcam/) so that your machine knows where to pull images from:

Author: Okay, will reword. I think the info about the models belongs (and is included) later.

### Configure the vision service

Now that you know the camera is properly connected to your machine, it is time to add computer vision by configuring the [vision service](/ml/vision/) on your machine.
Viam's built-in [`mlmodel` vision service](/ml/vision/mlmodel/) works with Tensor Flow Lite models, but since this tutorial uses a YOLOv8 model, we will use a {{< glossary_tooltip term_id="module" text="module" >}} from the [modular resource registry](/registry/) that augments Viam with YOLOv8 integration:

1. Navigate to the **Services** subtab of your machine's **Config** tab.

2. Click **Create service**.
   Start typing `yolo` and select **vision / yolov8** from the registry options.
   Click **Add module**.
   You can find more details about this module in [its readme](https://github.com/viam-labs/YOLOv8).

3. Give your vision service a name, for example `yolo`, and click **Create**.

4. In the **Attributes** box of your new vision service, paste the following JSON:

   ```json {class="line-numbers linkable-line-numbers"}
   {
     "model_location": "keremberke/yolov8n-hard-hat-detection"
   }
   ```

>    This tells the vision service where to look for [the hard hat detection model](https://huggingface.co/keremberke/yolov8s-hard-hat-detection) we are using for this tutorial.

Author: Okay, I'm adding this to the opening paragraph of this section.

Now you will pull them together with the [`objectfilter`](https://app.viam.com/module/felixreichenbach/object-filter) {{< glossary_tooltip term_id="module" text="module" >}}.
This module takes a vision service (in this case, your hard hat detector) and applies it to your webcam feed.
It outputs a stream with bounding boxes around the hard hats (and people without hard hats) in your camera's view so that you can see the detector working.
This module also filters the output so that later, when you configure data management, you can save only the images that contain people without hard hats.

1. Navigate to the **Components** subtab of your machine's **Config** tab.

2. Click **Create component**.
   Start typing `objectfilter` and select **camera / objectfilter** from the results.

3. Name your filtering camera something like `objectfilter-cam` and click **Create**.

4. Paste the following into the attributes box:

   ```json {class="line-numbers linkable-line-numbers"}
   {
     "filter_data": true,
     "camera": "my_webcam",
     "vision_services": ["yolo"],
     "labels": ["Hardhat", "NO-Hardhat"],
     "confidence": 0.5,
     "display_boxes": true
   }
   ```

>    If you named your detector something other than "yolo," edit the `detector_name` line accordingly.

Author: Oops thanks for catching. I switched from a different filter camera module to this one with different attribute names.
```suggestion
   If you named your detector something other than "yolo," edit the `vision_services` value accordingly.
```

   You can also edit the confidence threshold.

>    If you change it to `0.2` for example, you will see detections of over 20% confidence.

Author: Okay, rewording this sentence more descriptively.

>    If you change it to `0.2` for example, you will see detections of over 20% confidence.

Reviewer: thanks - this is great now

>    Setting the `filter_data` attribute to `true` means that later, when you configure data capture on this camera, only images that contain one or more of the labels will be captured.

Reviewer: ```suggestion
   Setting the `filter_data` attribute to `true` means that later, when you configure data capture on this camera, only images that have one or more of the labels will be captured and sent to the cloud.
   For testing purposes, leave both labels in the array for now.
   We will remove the `"Hardhat"` label from the configuration later so that the camera only captures and saves images when someone **isn't** wearing a hard hat.
```

I made a note here first asking why we were doing this because it would capture all the images. So good to explain it's for testing

   {{<imgproc src="/tutorials/helmet/filtercam-config.png" resize="x1100" declaredimensions=true alt="The detector_cam config panel in the Viam app." >}}

5. Click **Save config** at the bottom of the screen.

### Test the detector

Now that the detector is configured, it's time to test it!

1. Navigate to the **Control** tab.

2. Click the **objectfilter_cam** panel to open your detector camera controls.

3. Toggle **View objectfilter_cam** to the "on" position.
   This displays a live feed from your webcam with detection bounding boxes overlaid on it.

4. The detector can recognize humans with and without hard hats on.
   Try positioning yourself in front of the camera without anything on your head, and then with a helmet on, and watch the bounding boxes appear.

   {{<imgproc src="/tutorials/helmet/no-hard-hat1.png" resize="x1100" declaredimensions=true alt="A person with no hard hat on, with a bounding box labeled No-Hardhat around her head." >}}

## Configure data capture and sync

Viam's built-in [data management service](/data/) allows you to, among other things, capture images and sync them to the cloud.

> Configure data capture on the `objectfilter` camera to capture images of people without hard hats:

Reviewer: Or somethign like that, we want to explain why people would want to capture this
```suggestion
Configure data capture on the `objectfilter` camera to capture images of people without hard hats, so that you can see who was not wearing a hard hat:
```

Author: ```suggestion
For this project, you will capture images of people without hard hats so that you can see who wasn't wearing one, and so that you can trigger notifications when these images are captured and synced.
Configure data capture on the `objectfilter` camera to capture images of people without hard hats:
```

Reviewer: I like it!1

1. First, you need to add the data service to your machine to make it available to capture data on your camera.

   Navigate to the **Services** subtab of your machine's **Config** tab.

   Click **Create service**.
   Type "data" and click **data management / RDK**.
   Name your data management service `data-manager` and click **Create**.

   Leave all the default data service attributes as they are and click **Save config**.

2. Now you're ready to enable data capture on your detector camera.
   Navigate to the **Components** subtab.
   Locate the `objectfilter-cam` panel.

3. Click **Add method**.
   Click the **Type** dropdown and select **ReadImage**.
   Set the capture frequency to `0.2` images per second (equivalent to one image every 5 seconds).
   You can always change the frequency to suit your use case.

   {{<imgproc src="/tutorials/helmet/datacapture-config.png" resize="x1100" declaredimensions=true alt="The objectfilter cam configured with data capture set to read images at 0.2 hertz. Mime type is left at the default image/jpeg." >}}

### Test data capture and sync

To make sure the detector camera is capturing and syncing labeled images:

> 1. Position yourself in front of your webcam for 30 seconds or so to let it capture a few images of a person without a hard hat on.

Reviewer: ```suggestion
1. Position yourself in front of your webcam for approximately 30 seconds to let it capture a few images of a person without a hard hat on.
```

2. Navigate to your [**DATA** page](https://app.viam.com/data/view?view=images) in the Viam app.
   You should see some images with bounding boxes on them.
   If you do not, try refreshing the page.

   {{<imgproc src="/tutorials/helmet/synced-data.png" resize="x1000" declaredimensions=true alt="The data manager page in the Viam app, displaying three images of a person with a bounding box labeled NO-Hardhat around her face." >}}

3. You can also try this with a hard hat on your head.

>    {{<imgproc src="/tutorials/helmet/hardhat-data.png" resize="x1000" declaredimensions=true alt="The data manager page in the Viam app, displaying three images of a person with a bounding box labeled NO-Hardhat around her face." >}}

Reviewer: see above - want to check you're using images without copyright

Author: Yep done!

### Modify the detector config to sync only images without hard hats

Until now, you've been identifying people without hard hats as well as people with hard hats.
Now that you have verified that the detector and data sync are working, modify your config so that only images with people _without_ hard hats are captured:

1. Navigate to your `objectfilter-cam` card on the **Config** tab.

2. Delete the `"Hardhat",` line from the `"labels"` array.

3. Click **Save config**.

## Set up email notifications

> [Webhooks](https://en.wikipedia.org/wiki/Webhook) are a way to trigger a custom function to run when a certain event happens.
> For the purposes of this project, you will use a webhook to run a function that sends you an email when an image of someone without a hard hat is uploaded to the cloud.

Reviewer: Let's link to our own docs. Webhooks I think are a request, it's not necessarily a function that gets executed. I've 
```suggestion
[Webhooks](/build/configure/#webhooks) allow you to trigger actions by sending an HTML request when a certain event happens.
In this case, you're going to set up a webhook to trigger a serverless function that sends you an email when an image of someone without a hard hat is uploaded to the cloud.
```

### Create a serverless function

A serverless function is a simple script that is hosted by a service such as [Google Cloud Functions](https://cloud.google.com/functions) or [AWS Lambda](https://aws.amazon.com/pm/lambda).
You don't need to host it on your machine; instead, it is always available and runs only when an event triggers it.

#### Set up your cloud function scaffold

For this project, we used Google Cloud Functions, and wrote our function in Python.
If you are new to cloud functions, you may find [this getting started guide](https://cloud.google.com/functions/docs/console-quickstart) useful.

1. Create a Google Cloud (GCP) account.
2. Create a project, then search for **Cloud Functions** in the Google Cloud console search bar.
3. Click **CREATE FUNCTION**.
4. Choose `1st gen` as the **Environment**.
5. Give your function a name.
   We called ours "hat-email."
6. Choose your region.
7. For trigger type, choose `HTTP`.
8. Click **NEXT**.
9. For **Runtime**, choose `Python 3.8`.
10. Click **Deploy**.
    The demo function won't do anything exciting yet; keep going through the next few sections before testing it.

#### Configure email credentials

> To write a function that sends an email, you need to configure credentials to authenticate to the sending email account.
> For this project we used [SendGrid](https://sendgrid.com) (which has a free tier) to make configuration of your email notifications simpler:

Reviewer: You want to introduce sendgrid first. You don't need credentials, you need something that can send emails. That service happens to have an integration for google cloud functions and that needs credentials. Bit putting the cart before the horse ;)
```suggestion
To write a function that sends an email, you need a service that can send emails.
For this project we used [SendGrid](https://sendgrid.com) (which has a free tier) to make configuration of your email notifications simpler.
Follow these instructiong to create a sendgrid account and configure the SendGrid API credentials for your Google Cloud Function:
```

Author: ```suggestion
To write a function that sends an email, you need a service that can send emails.
For this project we used [SendGrid](https://sendgrid.com) (which has a free tier) to make configuration of your email notifications simpler.
Follow these instructions to create a SendGrid account and configure the SendGrid API credentials for your Google Cloud Function:
```

1. [Enable SendGrid Email API](https://console.cloud.google.com/marketplace/details/sendgrid-app/sendgrid-email) through the Google Cloud Marketplace.
1. Create a SendGrid account and confirm your email.
1. Create a [SendGrid API key](https://docs.sendgrid.com/ui/account-and-settings/api-keys) with **Mail Send > Full Access** permissions.
1. Store your API key with the Google Cloud Function console:
   1. In the top banner of your console, click **Edit**.
   2. Click to expand **Runtime, build, connections and security settings**.
   3. Under **Runtime environment variables**, click **ADD VARIABLE**.
      For the name, enter `EMAIL_API_KEY`.
      For the value, paste the API key you generated with SendGrid.
   4. Click **NEXT**.

#### Write the cloud function code

SendGrid's [Email API Quickstart](https://docs.sendgrid.com/for-developers/sending-email/quickstart-python) contains more information about the functionality of the SendGrid API.
The following code is adapted from that example.

> 1. Copy and paste the following code into your cloud function source code `main.py` file:

Reviewer: ```suggestion
1. Copy and paste the following code into your cloud function source code `main.py` file and change the email address parameters `from_email` and `to_emails`:
```

   ```python {class="line-numbers linkable-line-numbers"}
   import functions_framework
   import os
   import sendgrid
   from sendgrid import SendGridAPIClient
   from sendgrid.helpers.mail import Mail, Email, To, Content
   from python_http_client.exceptions import HTTPError

   def email(request):

       sg = sendgrid.SendGridAPIClient(api_key=os.environ.get('EMAIL_API_KEY'))

       message = Mail(

>         # Change to the email you created the API key for
>         from_email='youremailaddresstosendfrom@example.com',
>         # Change to whichever email should receive the notification
>         to_emails='youremailaddresstosendto@example.com',

Reviewer: Highlighting the call to action a bit more
```suggestion
        # TODO: Change to the email you created the API key for
        from_email='youremailaddresstosendfrom@example.com',
        # TODO: Change to whichever email should receive the notification
        to_emails='youremailaddresstosendto@example.com',
```

#### Test the cloud function

Now you can test the script:

1. Navigate to the **TRIGGER** tab in the Google Cloud console.

2. Click the copy button to copy the trigger URL.
   Paste it into a new browser window and hit your **Enter** key.
   The browser should display "Notification sent."
   If it displays an error, double-check your code, and double-check your [access authorization](https://cloud.google.com/functions/docs/securing/managing-access-iam).

3. Check your email inbox.

   {{<imgproc src="/tutorials/helmet/email-received.png" resize="x600" declaredimensions=true alt="The email, opened in a web email client." style="max-width: 400px" >}}

   If you don't see an email, check your spam folder.
   If you still don't see an email, make sure your SendGrid account is fully set up (2FA enabled, email confirmed) and that your email API key is [correctly configured](#configure-email-credentials).

### Configure a webhook on your machine

Now it's time to configure a [webhook](/build/configure/#webhooks) on your machine to trigger the email cloud function when a person is not wearing a hard hat.

> Since you configured data to sync only when an image of a person without a hard hat is captured, configuring the webhook to trigger each time an image is synced to the cloud will produce the desired result.

Reviewer: nice

Configure a webhook as follows:

1. Navigate to the **Config** tab of your machine.
   Toggle the **Mode** to **Raw JSON**.

2. Paste the following JSON template into your raw JSON config.
   `"webhooks"` is a top-level section like `"components"`, `"services"`, or any of the other config sections.

   ```json {class="line-numbers linkable-line-numbers"}
     "webhooks": [
       {
         "url": "<Insert your own cloud function URL>",
         "event": {
           "attributes": {
             "data_types": ["binary"]
           },
           "type": "part_data_ingested"
         }
       }
     ]
   ```

3. Replace the `url` value with your cloud function URL.
   You can get this URL by copying it from the **TRIGGER** tab in the cloud function console.

>    Once you've done this, the `url` line should resemble, for example, `"url": "https://us-east1-blah-blah-123456.cloudfunctions.net/hat-email"`.

Reviewer: ```suggestion
   Once you've done this, the `url` line should resemble `"url": "https://us-east1-abcd-efgh-123456.cloudfunctions.net/hat-email"`.
```

Author: I included "for example" because sometimes when we say something should resemble something, we really mean it should be an exact match. Fine to replace the blah blah part but that part is actually words, not random characters, so would this be okay?
```suggestion
   Once you've done this, the `url` line should resemble, for example, `"url": "https://us-east1-example-string-123456.cloudfunctions.net/hat-email"`.
```

Reviewer: yep sounds good - just don't like the "blah" "blah". That's maybe a me-thing but it feels a bit unprofessional 🙈

Author: That's fair XD

4. Click **Save config**.

## Test the whole system

You've built all the pieces of the system and connected them together.
Now it's time to test the whole thing.

Make sure `viam-server` is running on your machine.
Position yourself, without a hard hat, in front of your camera.

> Wait a couple of minutes for the email to arrive in your inbox, then celebrate your success when it does!
> Great work!

Reviewer: I think this is a little too many exclamation marks
```suggestion
Wait a couple of minutes for the email to arrive in your inbox. Congratulations, you've successfully built your hard hat monitor!
```

Author: ```suggestion
Wait a couple of minutes for the email to arrive in your inbox.
Congratulations, you've successfully built your hard hat monitor!
```

## Next steps

Here are some ways you could expand on this project:

- Mount a camera on a rover and either drive it manually using remote control, or use the motion and navigation services to plan paths for the rover.
  Or, mount cameras in multiple places.

- Change your cloud function to send a different kind of notification, or trigger some other action.
  For an example demonstrating how to configure text notifications, see the [Detect a Person and Send a Photo tutorial](/tutorials/projects/send-security-photo/).

> - Use a different existing model or [train your own](/ml/train-model/), to detect and send notifications about something else such as [forklifts](https://huggingface.co/keremberke/yolov8m-forklift-detection).

Reviewer: ```suggestion
- Use a different existing model or [train your own](/ml/train-model/), to detect and send notifications about something else such as [forklifts](https://huggingface.co/keremberke/yolov8m-forklift-detection) appearing in your camera stream.
```

{{< cards >}}
{{% card link="/tutorials/projects/send-security-photo/" %}}
{{% card link="/ml/train-model/" %}}

> {{% card link="/tutorials/configure/scuttlebot/" %}}

Reviewer: I think that's less relevant
```suggestion
```

Author: I was thinking because they could put their camera on a rover? Do you think the navigation one is enough for rovers?

Reviewer: I think adding too many links could be overwhelming, you could use the scutle one instead of the navigate one if you want.
